RecoMET/METAlgorithms/src/HcalHaloAlgo.cc

```diff
@@ -37,100 +36,92 @@ HcalHaloData HcalHaloAlgo::Calculate(const CaloGeometry& TheCaloGeometry, edm::H
-	{ 
-	  SumE[iPhi]+= hit->energy();
-	  NumHits[iPhi] ++;
+  for (const auto & hit : (HBHERecHitCollection)(*TheHBHERecHits)) {
+    HcalDetId id = HcalDetId(hit.id());
+    switch ( id.subdet() ) {      
+    case HcalBarrel:
+      if(hit.energy() < HBRecHitEnergyThreshold )continue;
+      break;
+    case HcalEndcap:
+      if(hit.energy() < HERecHitEnergyThreshold ) continue;
+      break;
+    default:
+      continue;
+    }
+    
+    int iEta = id.ieta();
+    int iPhi = id.iphi();
+    if(iPhi < nPhiMax && std::abs(iEta) <= iEtaOverlap) { 
+      SumE[iPhi]+= hit.energy();
+      NumHits[iPhi] ++;
 	  
-	  float time = hit->time();
-	  MinTimeHits[iPhi] = time < MinTimeHits[iPhi] ? time : MinTimeHits[iPhi];
-	  MaxTimeHits[iPhi] = time > MaxTimeHits[iPhi] ? time : MaxTimeHits[iPhi];
-	}
+      float time = hit.time();
+      MinTimeHits[iPhi] = time < MinTimeHits[iPhi] ? time : MinTimeHits[iPhi];
+      MaxTimeHits[iPhi] = time > MaxTimeHits[iPhi] ? time : MaxTimeHits[iPhi];
     }
+  }
   
-  for( int iPhi = 1 ; iPhi < 73 ; iPhi++ )
-    {
-      if( SumE[iPhi] >= SumEnergyThreshold && NumHits[iPhi] > NHitsThreshold )
-	{
-	  // Build PhiWedge and store to HcalHaloData if energy or #hits pass thresholds
-	  PhiWedge wedge(SumE[iPhi], iPhi, NumHits[iPhi], MinTimeHits[iPhi], MaxTimeHits[iPhi]);
+  for (int iPhi = 1 ; iPhi < nPhiMax ; iPhi++ )   {
+    if( SumE[iPhi] >= SumEnergyThreshold && NumHits[iPhi] > NHitsThreshold ) {
+      // Build PhiWedge and store to HcalHaloData if energy or #hits pass thresholds
+      PhiWedge wedge(SumE[iPhi], iPhi, NumHits[iPhi], MinTimeHits[iPhi], MaxTimeHits[iPhi]);
 	  
-	  // Loop over rechits again to calculate direction based on timing info
-	  std::vector<const HBHERecHit*> Hits;
-	  for( HBHERecHitCollection::const_iterator hit = TheHBHERecHits->begin() ; hit != TheHBHERecHits->end() ; hit++ )
-	    {
-
-	      HcalDetId id = HcalDetId(hit->id());
-	      if( id.iphi() != iPhi ) continue;
-	      if( TMath::Abs(id.ieta() ) > 22 ) continue;  // has to overlap geometrically w/ HB
-	      switch ( id.subdet() )
-		{
-		case HcalBarrel:
-		  if(hit->energy() < HBRecHitEnergyThreshold )continue;
-		  break;
-		case HcalEndcap:
-		  if(hit->energy() < HERecHitEnergyThreshold ) continue;
-		  break;
-		default:
-		  continue;
-		}
-	      Hits.push_back(&(*hit));
-	    }
+      // Loop over rechits again to calculate direction based on timing info
+      std::vector<const HBHERecHit*> Hits;
+      for (const auto & hit : (HBHERecHitCollection)(*TheHBHERecHits)) {
+	HcalDetId id = HcalDetId(hit.id());
+	if( id.iphi() != iPhi ) continue;
+	if( std::abs(id.ieta() ) > iEtaOverlap ) continue;  // has to overlap geometrically w/ HB
+	switch ( id.subdet() ) {
+	case HcalBarrel:
+	  if(hit.energy() < HBRecHitEnergyThreshold )continue;
+	  break;
+	case HcalEndcap:
+	  if(hit.energy() < HERecHitEnergyThreshold ) continue;
+	  break;
+	default:
+	  continue;
+	}
+	Hits.push_back(&(hit));
```

Comment:
I think this is loop is the origin of the problem. The cast done in the loop creates a temporary container. Then the address of an item in that container is stored in `Hits`. Then when we leave the loop, the temporary container goes out of scope so all the pointers in `Hits` are pointing to invalid memory.